network/p2p/scoring/registry_test.go

```diff
@@ -385,6 +385,112 @@ func testScoreRegistrySpamRecordWithSubscriptionPenalty(t *testing.T, messageTyp
+// The function focuses on evaluating the registry's response to spam activities (as represented by control messages) from a
+// peer that has sent duplicate messages. It verifies that penalties are accurately computed and applied, taking into account both
+// the spam record and the duplicate message's penalty.
+func testScoreRegistrySpamRecordWithDuplicateMessagesPenalty(t *testing.T, messageType p2pmsg.ControlMessageType, expectedPenalty float64) {
+	peerID := unittest.PeerIdFixture(t)
+	cfg, err := config.DefaultConfig()
+	require.NoError(t, err)
+	// refresh cached app-specific score every 100 milliseconds to speed up the test.
+	cfg.NetworkConfig.GossipSub.ScoringParameters.AppSpecificScore.ScoreTTL = 10 * time.Millisecond
+
+	duplicateMessagesCount := 100.0
+	reg, spamRecords, appScoreCache := newGossipSubAppSpecificScoreRegistry(t, cfg.NetworkConfig.GossipSub.ScoringParameters,
+		withStakedIdentities(peerID),
+		withValidSubscriptions(peerID),
+		func(registryConfig *scoring.GossipSubAppSpecificScoreRegistryConfig) {
+			registryConfig.GetDuplicateMessageCount = func(_ peer.ID) float64 {
+				return duplicateMessagesCount
+			}
+		})
+
+	// starts the registry.
+	ctx, cancel := context.WithCancel(context.Background())
+	signalerCtx := irrecoverable.NewMockSignalerContext(t, ctx)
+	reg.Start(signalerCtx)
+	unittest.RequireCloseBefore(t, reg.Ready(), 1*time.Second, "failed to start GossipSubAppSpecificScoreRegistry")
+
+	// initially, the spamRecords should not have the peer id; also the app specific score record should not be in the cache.
+	require.False(t, spamRecords.Has(peerID))
+	score, updated, exists := appScoreCache.Get(peerID) // get the score from the cache.
+	require.False(t, exists)
+	require.Equal(t, time.Time{}, updated)
+	require.Equal(t, float64(0), score)
+
+	expectedDuplicateMessagesPenalty := duplicateMessagesCount * scoring.DefaultDuplicateMessagePenalty
+
+	// eventually, the app specific score should be updated in the cache.
+	require.Eventually(t, func() bool {
+		// calling the app specific score function when there is no app specific score in the cache should eventually update the cache.
+		score := reg.AppSpecificScoreFunc()(peerID)
+		// since the peer id does not have a spam record, the app specific score should be the max app specific reward minus
+		// the expected penalty for 100 duplicate messages
+		return scoring.MaxAppSpecificReward-score == math.Abs(expectedDuplicateMessagesPenalty)
+	}, 5*time.Second, 100*time.Millisecond)
+
+	// report a misbehavior for the peer id.
+	reg.OnInvalidControlMessageNotification(&p2p.InvCtrlMsgNotif{
+		PeerID:  peerID,
+		MsgType: messageType,
+	})
+
+	// the penalty should now be updated in the spamRecords
+	record, err, ok := spamRecords.Get(peerID) // get the record from the spamRecords.
+	assert.True(t, ok)
+	assert.NoError(t, err)
+	assert.Less(t, math.Abs(expectedPenalty-record.Penalty), 10e-3)        // penalty should be updated to -10.
```

Comment:
Please use `unittest.RequireNumericallyClose` instead, for this and other repetitions of this.

Reply:
https://github.com/onflow/flow-go/pull/5058/commits/ee5d747ed7b2d21c69055b10e10bfa151124a5ea